### PR TITLE
feat(frontend/shared): Update the design of the "Chats" page tm-388

### PR DESCRIPTION
**Pull request by GEOFARL**
image
image
image
image
image
image

### CODE REVIEW

File: apps/frontend/package.json [hunk 1]
```diff
@@ -21,11 +21,13 @@
 		"@reduxjs/toolkit": "2.0.1",
 		"@trackmates/shared": "*",
 		"clsx": "2.1.0",
+		"date-fns": "3.3.1",
```
Comment (Ann-Maksay) [inline apps/frontend/package.json:24]: I think all helpers that use `date-fns` should be in `shared`, we already have that package there

Comment (GEOFARL) [reply apps/frontend/package.json:24]: But does it make sense to put functions in the shared folder that will only be used on the frontend?

Comment (Ann-Maksay) [reply apps/frontend/package.json:24]: I guess [this](https://github.com/BinaryStudioAcademy/bsa-winter-2023-2024-trackmates/pull/267#discussion_r1508827432) is also only used on the frontend, so it doesn't really matter

Comment (GEOFARL) [reply apps/frontend/package.json:24]: Thanks, i'll move it to shared

File: apps/frontend/src/pages/chats/libs/components/chat/chat.tsx [hunk 2]
```diff
@@ -1,9 +1,16 @@
+import { Fragment } from "react";
+
 import defaultAvatar from "~/assets/img/default-avatar.png";
-import { Image, Loader } from "~/libs/components/components.js";
+import { Button, Image, Loader } from "~/libs/components/components.js";
+import { LAST_ARRAY_ITEM } from "~/libs/constants/constants.js";
+import { checkIsDatePrecedesAnotherByOneDay } from "~/libs/helpers/helpers.js";
+import { useAppDispatch, useCallback } from "~/libs/hooks/hooks.js";
 import { type ChatMessageItemResponseDto } from "~/modules/chat-messages/chat-messages.js";
+import { actions } from "~/modules/chats/chats.js";
```
Comment (what1s1ove) [inline apps/frontend/src/pages/chats/libs/components/chat/chat.tsx:9]: rename

File: apps/frontend/package.json [hunk 3]
```diff
@@ -26,6 +26,7 @@
 		"react-hook-form": "7.49.3",
 		"react-redux": "9.1.0",
 		"react-router-dom": "6.21.3",
+		"react-textarea-autosize": "8.5.3",
```
Comment (what1s1ove) [inline apps/frontend/package.json:29]:
```suggestion
```
use just a html tag

File: apps/frontend/src/libs/components/textarea/textarea.tsx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+const Textarea = <T extends FieldValues>({
+	className,
+	control,
+	errors,
+	hasVisuallyHiddenLabel,
+	label,
+	name,
+	placeholder = "",
+	rows,
```
Comment (what1s1ove) [inline apps/frontend/src/libs/components/textarea/textarea.tsx:33]: you do not have this component. you need to add a `rows` prop to the `Input` component. And if this props is passed you need to render a `textarea`

File: packages/shared/src/libs/helpers/get-formatted-chat-date/get-formatted-chat-date.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import { format, isThisYear, isToday, isYesterday } from "date-fns";
+
+const getFormattedChatDate = (dateString: string): string => {
```
Comment (what1s1ove) [inline packages/shared/src/libs/helpers/get-formatted-chat-date/get-formatted-chat-date.ts:3]:
```suggestion
```
you need to use get formatted date instead. this is a frontend helper only, so move it there

File: apps/frontend/src/pages/chats/libs/components/chat/chat.tsx [hunk 6]
```diff
@@ -21,28 +28,77 @@ const Chat: React.FC<Properties> = ({
...
-						isCurrentUserSender={receiver.id !== message.senderUser.id}
-						key={message.id}
-						message={message}
+				{messages.map((message) => {
+					if (checkIsDatePrecedesAnotherByOneDay(message.createdAt, lastDate)) {
+						const previousDate = lastDate;
+						lastDate = message.createdAt;
+
+						return (
+							<Fragment key={message.id}>
```
Comment (what1s1ove) [inline apps/frontend/src/pages/chats/libs/components/chat/chat.tsx:74]: create a new component for message item

File: apps/frontend/src/pages/chats/libs/components/chat/chat.tsx [hunk 7]
```diff
@@ -21,28 +28,77 @@ const Chat: React.FC<Properties> = ({
 	onSubmit,
 	receiver,
 }: Properties) => {
+	const dispatch = useAppDispatch();
+
+	const handleClick = useCallback((): void => {
+		dispatch(actions.leaveChat());
+	}, [dispatch]);
+
+	let lastDate = "";
```
Comment (what1s1ove) [inline apps/frontend/src/pages/chats/libs/components/chat/chat.tsx:37]:
```suggestion
```
bad solution. try to find a better solution

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: packages/shared/src/libs/helpers/get-formatted-chat-date/get-formatted-chat-date.ts (hunk 5 above)
Comment (GEOFARL) [reply packages/shared/src/libs/helpers/get-formatted-chat-date/get-formatted-chat-date.ts:3]: The getFormattedDate fn that we already have right now does not produce the result that we need to have in the chat window, needed to create a separate helper. Moved it to the shared folder, to not install 'date-fns' package on the frontend side as was pointed out [here](https://github.com/BinaryStudioAcademy/bsa-winter-2023-2024-trackmates/pull/407#discussion_r1516854005)

File: apps/frontend/package.json (hunk 3 above)
Comment (GEOFARL) [reply apps/frontend/package.json:29]: image
It just makes it possible for native textarea element to grow according to the amount of content

File: apps/frontend/src/pages/chats/libs/components/chat/chat.tsx (hunk 6 above)
Comment (GEOFARL) [reply apps/frontend/src/pages/chats/libs/components/chat/chat.tsx:74]: We already have a component for message `ChatMessage`. In this case `ChatDate` is used to add date delimiter when a message starts from a new date
image

Do not think that it makes sense to create a new component out of it

File: packages/shared/src/libs/helpers/get-formatted-chat-date/get-formatted-chat-date.ts (hunk 5 above)
Comment (Ann-Maksay) [reply packages/shared/src/libs/helpers/get-formatted-chat-date/get-formatted-chat-date.ts:3]: What is meant here is that you should leave all date-fsn-related functions in shared in wrappers and combine them in getFormattedChatDate on the frontend.  Take a look at the getFormattedChatDate helper in shared.
Sorry if I didn't explain it clearly enough in my comment

File: apps/frontend/package.json (hunk 3 above)
Comment (what1s1ove) [reply apps/frontend/package.json:29]: give the static height. it will be enough

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: apps/frontend/src/libs/components/input/input.tsx [hunk 8]
```diff
@@ -59,16 +66,36 @@ const Input = <T extends FieldValues>({
 		hasVisuallyHiddenLabel && "visually-hidden",
 	);
 
+	const handleKeyDown = useCallback(
```
Comment (what1s1ove) [inline apps/frontend/src/libs/components/input/input.tsx:69]:
```suggestion
```
without any specific behaviour

File: apps/frontend/src/libs/components/input/input.tsx [hunk 9]
```diff
@@ -21,7 +22,9 @@ type Properties<T extends FieldValues> = {
 	iconName?: IconName;
 	label: string;
 	name: FieldPath<T>;
+	onSubmit?: (_event: React.BaseSyntheticEvent) => void;
```
Comment (what1s1ove) [inline apps/frontend/src/libs/components/input/input.tsx:25]:
```suggestion
```
textereat shouldn't know anything about submit

File: apps/frontend/src/libs/components/input/styles.module.css [hunk 10]
```diff
@@ -22,12 +22,27 @@
...
 
-.input::placeholder {
+.input::placeholder,
+.textarea::placeholder {
 	font-weight: 400;
 	color: var(--color-menu);
 	opacity: 0.6;
 }
 
+.textarea {
```
Comment (what1s1ove) [inline apps/frontend/src/libs/components/input/styles.module.css:32]: you can use the same class name as for the input

File: apps/frontend/src/libs/components/input/input.tsx (hunk 8 above)
Comment (what1s1ove) [reply apps/frontend/src/libs/components/input/input.tsx:69]: you do not need to handle it at all this time

File: packages/shared/src/libs/helpers/helpers.ts [hunk 11]
```diff
@@ -6,3 +7,4 @@ export { getSanitizedHtml } from "./get-sanitized-html/get-sanitized-html.helper
 export { getSizeInBytes } from "./get-size-in-bytes/get-size-in-bytes.helper.js";
 export { getTimeDistanceFormatDate } from "./get-time-distance-format-date/get-time-distance-format-date.helper.js";
 export { initDebounce } from "./init-debounce/init-debounce.helper.js";
+export { isThisYear, isToday, isYesterday } from "date-fns";
```
Comment (what1s1ove) [inline packages/shared/src/libs/helpers/helpers.ts:10]:
```suggestion
```
create yourown wrappers for these functions

File: packages/shared/src/libs/constants/constants.ts [hunk 12]
```diff
@@ -1,2 +1,4 @@
 export { EMPTY_ARRAY_LENGTH } from "./empty-array-length.constant.js";
+export { LAST_ARRAY_ITEM } from "./last-array-item.js";
```
Comment (YourTeaChan) [inline packages/shared/src/libs/constants/constants.ts:2]:
```suggestion
export { LAST_ARRAY_ITEM } from "./last-array-item.constant.js";
```

File: packages/shared/src/libs/constants/constants.ts [hunk 13]
```diff
@@ -1,2 +1,4 @@
...
 export { MAX_FILE_SIZE_IN_MB } from "./max-file-size.constant.js";
+export { PREVIOUS_INDEX_OFFSET } from "./previous-index-offset.js";
```
Comment (YourTeaChan) [inline packages/shared/src/libs/constants/constants.ts:4]:
```suggestion
export { PREVIOUS_INDEX_OFFSET } from "./previous-index-offset.constant.js";
```

File: packages/shared/src/libs/helpers/helpers.ts [hunk 14]
```diff
@@ -1,3 +1,4 @@
+export { checkIsDatePrecedesAnotherByOneDay } from "./check-is-date-precedes-another-by-one-day/check-is-date-precedes-another-by-one-day.js";
```
Comment (YourTeaChan) [inline packages/shared/src/libs/helpers/helpers.ts:1]:
```suggestion
export { checkIsDatePrecedesAnotherByOneDay } from "./check-is-date-precedes-another-by-one-day/check-is-date-precedes-another-by-one-day.helper.js";
```

File: apps/frontend/src/pages/chats/libs/components/chat/chat.tsx [hunk 15]
```diff
@@ -21,28 +29,78 @@ const Chat: React.FC<Properties> = ({
...
 			</div>
 			<ul className={styles["chat-container"]}>
-				{messages.map((message) => (
-					<ChatMessage
-						isCurrentUserSender={receiver.id !== message.senderUser.id}
-						key={message.id}
-						message={message}
+				{messages.flatMap((message, index) => {
+					const components = [];
+					const previousMessage = messages[index - PREVIOUS_INDEX_OFFSET];
```
Comment (what1s1ove) [inline apps/frontend/src/pages/chats/libs/components/chat/chat.tsx:69]: it's really strange logic here. why we can't just map everything?

in render we should not have such logic

move calculation of messages (not components renders) to helper and then use it here

File: apps/frontend/src/pages/chats/libs/components/chat/chat.tsx [hunk 16]
```diff
@@ -1,9 +1,17 @@
 import defaultAvatar from "~/assets/img/default-avatar.png";
-import { Image, Loader } from "~/libs/components/components.js";
+import { Button, Image, Loader } from "~/libs/components/components.js";
+import {
+	LAST_ARRAY_ITEM,
+	PREVIOUS_INDEX_OFFSET,
+} from "~/libs/constants/constants.js";
+import { AppRoute } from "~/libs/enums/app-route.enum.js";
```
Comment (what1s1ove) [inline apps/frontend/src/pages/chats/libs/components/chat/chat.tsx:7]: import

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: packages/shared/src/libs/helpers/helpers.ts (hunk 11 above)
Comment (fshabanov) [inline packages/shared/src/libs/helpers/helpers.ts:10]: Rename to be compatible with B20

File: apps/frontend/src/pages/chats/libs/components/chat-date/chat-date.tsx [hunk 17]
```diff
@@ -0,0 +1,19 @@
+import { getFormattedChatDate } from "../../../libs/helpers/helpers.js";
```
Comment (fshabanov) [inline apps/frontend/src/pages/chats/libs/components/chat-date/chat-date.tsx:1]: Fix import path

File: apps/frontend/src/pages/chats/libs/components/chat/chat.tsx (hunk 16 above)
Comment (fshabanov) [inline apps/frontend/src/pages/chats/libs/components/chat/chat.tsx:7]:
```suggestion
import { AppRoute } from "~/libs/enums/enums.js";
```

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/libs/components/input/input.tsx [hunk 18]
```diff
@@ -21,7 +21,9 @@ type Properties<T extends FieldValues> = {
 	iconName?: IconName;
 	label: string;
 	name: FieldPath<T>;
+	onKeyDown?: (_event: React.KeyboardEvent) => void;
```
Comment (what1s1ove) [inline apps/frontend/src/libs/components/input/input.tsx:24]:
```suggestion
```
you do not to handle any specific keys. remove this handler from common components

File: apps/frontend/src/libs/enums/event-code.enum.ts [hunk 19]
```diff
@@ -0,0 +1,5 @@
+const EventCode = {
```
Comment (what1s1ove) [inline apps/frontend/src/libs/enums/event-code.enum.ts:1]:
```suggestion
```
do not handle this this time

File: apps/frontend/src/pages/chats/libs/helpers/helpers.ts [hunk 20]
```diff
@@ -0,0 +1,2 @@
+export { getFormattedChatDate } from "./get-formatted-chat-date.helper.js";
```
Comment (what1s1ove) [inline apps/frontend/src/pages/chats/libs/helpers/helpers.ts:1]: each helper should be in a separate folder

**Review by what1s1ove — CHANGES_REQUESTED** [review]

**Review by what1s1ove — APPROVED** [review]
@GEOFARL resolve comments please

File: apps/frontend/src/pages/chats/libs/helpers/get-formatted-chat-date/get-formatted-chat-date.helper.ts [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+	} else if (checkIsYesterday(date)) {
+		return DateValue.YESTERDAY;
+	} else if (checkIsThisYear(date)) {
+		return getFormattedDate(dateString, FormatDateType.MMMM_DO).toLowerCase();
+	} else {
+		return getFormattedDate(
+			dateString,
+			FormatDateType.MMMM_DO_YYYY,
+		).toLowerCase();
+	}
```
Comment (Oleksandra-Masiuk) [inline apps/frontend/src/pages/chats/libs/helpers/get-formatted-chat-date/get-formatted-chat-date.helper.ts:23]:
```suggestion
	if (checkIsToday(date)) {
		return DateValue.TODAY;
	} 
	if (checkIsYesterday(date)) {
		return DateValue.YESTERDAY;
	}
	if (checkIsThisYear(date)) {
		return getFormattedDate(dateString, FormatDateType.MMMM_DO).toLowerCase();
	} 
	return getFormattedDate(
			dateString,
			FormatDateType.MMMM_DO_YYYY,
		).toLowerCase();
```

File: apps/frontend/src/pages/chats/libs/helpers/prepare-message-items/prepare-message-items.helper.ts [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+const prepareMessageItems = (
+	messages: ChatMessageItemResponseDto[],
+): MessageItem[] => {
+	const items = [];
+
+	for (const [index, message] of messages.entries()) {
+		const previousMessage = messages[index - PREVIOUS_INDEX_OFFSET];
+
+		if (
+			previousMessage &&
```
Comment (Oleksandra-Masiuk) [inline apps/frontend/src/pages/chats/libs/helpers/prepare-message-items/prepare-message-items.helper.ts:17]: Could you please, move this condition to a separate variable with a descriptive name?

**Review by Oleksandra-Masiuk — CHANGES_REQUESTED** [review]

File: apps/frontend/src/pages/chats/libs/components/chat/chat.tsx [hunk 23]
```diff
@@ -21,28 +27,62 @@ const Chat: React.FC<Properties> = ({
...
+							alt="User avatar"
+							height="48"
+							shape="circle"
+							src={receiver.avatarUrl ?? defaultAvatar}
+							width="48"
+						/>
+					</div>
+					<span
+						className={styles["full-name"]}
+					>{`${receiver.firstName} ${receiver.lastName}`}</span>
```
Comment (fshabanov) [inline apps/frontend/src/pages/chats/libs/components/chat/chat.tsx:61]:
```suggestion
					>{receiver.firstName} {receiver.lastName}</span>
```

**Review by fshabanov — CHANGES_REQUESTED** [review]

**Review by what1s1ove — APPROVED** [review]

**Pull request merged**